common/ecdh.c

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright 2015-2018 Yubico AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+
+#ifdef _WIN32_BCRYPT
+#include <windows.h>
+#include <bcrypt.h>
+#else
+#include <openssl/evp.h>
+#include <openssl/ec.h>
+#include <openssl/ecdh.h>
+#include <openssl/rand.h>
+
+#include "openssl-compat.h"
+#endif
+
+#include "ecdh.h"
+
+#ifdef _WIN32_BCRYPT
+
+// TODO
+
+int ecdh_curve_p256(void) { return 0; }
+int ecdh_calculate_public_key(int curve, const uint8_t *privkey,
+                              size_t cb_privkey, uint8_t *pubkey,
+                              size_t cb_pubkey) {
+  return 0;
+}
+int ecdh_generate_keypair(int curve, uint8_t *privkey, size_t cb_privkey,
+                          uint8_t *pubkey, size_t cb_pubkey) {
+  return 0;
+}
+int ecdh_calculate_secret(int curve, const uint8_t *privkey, size_t cb_privkey,
+                          const uint8_t *pubkey, size_t cb_pubkey,
+                          uint8_t *secret, size_t cb_secret) {
+  return 0;
+}
+
+#else
+
+int ecdh_curve_p256(void) { return NID_X9_62_prime256v1; }
+
+int ecdh_calculate_public_key(int curve, const uint8_t *privkey,
+                              size_t cb_privkey, uint8_t *pubkey,
+                              size_t cb_pubkey) {
+  BN_CTX *ctx = BN_CTX_new();
+  BIGNUM *order = BN_new();
+  BIGNUM *pvt = BN_bin2bn(privkey, cb_privkey, NULL);
+  EC_GROUP *group = EC_GROUP_new_by_curve_name(curve);
+  if (ctx == NULL || order == NULL || pvt == NULL || group == NULL) {
+    EC_GROUP_free(group);
+    BN_free(pvt);
+    BN_free(order);
+    BN_CTX_free(ctx);
```

Comment:
a `goto` would make this more readable IMO.

Reply:
It's my allergy to goto...

Reply:
But fixed in 3ce940dca5a0a563b3bc836119e7eba43b6182c4